numpy/core/arrayprint.py

```diff
@@ -1203,20 +1206,36 @@ def array_repr(arr, max_line_width=None, precision=None, suppress_small=None):
 
     if skipdtype:
         return "%s(%s)" % (class_name, lst)
+
+    # determine typename
+    if issubclass(arr.dtype.type, flexible):
+        if arr.dtype.names:
+            typename = "%s" % str(arr.dtype)
+        else:
+            typename = "'%s'" % str(arr.dtype)
     else:
         typename = arr.dtype.name
-        # Quote typename in the output if it is "complex".
+        # quote typenames which can't be represented as python variable names
         if typename and not (typename[0].isalpha() and typename.isalnum()):
             typename = "'%s'" % typename
 
-        lf = ' '
+    prefix = "{}({},".format(class_name, lst)
+    suffix = "dtype={})".format(typename)
+
+    # compute whether we should put dtype on a new line: Do so if either
+    # the output already had multiple lines, or if adding the dtype
+    # would extend the line past max_line_width.
+    newline_ind = prefix.rfind('\n')
+    if _format_options['legacy'] == '1.13':
+        spacer = ' '
         if issubclass(arr.dtype.type, flexible):
-            if arr.dtype.names:
-                typename = "%s" % str(arr.dtype)
-            else:
-                typename = "'%s'" % str(arr.dtype)
-            lf = '\n'+' '*len(class_name + "(")
-        return "%s(%s,%sdtype=%s)" % (class_name, lst, lf, typename)
+            spacer = '\n' + ' '*len(class_name + "(")
+    elif newline_ind == -1 and len(prefix) + len(suffix) + 1 <= max_line_width:
+        spacer = ' '
```

Comment:
This isn't quite the rule I was proposing. I was also intending to keep the following:
```python
>>> np.zeros((3, 1), 'f2')
array([[0.],
       [0.],
       [0.]], dtype=float16)
```

So the test would be something like `max(len(l) for l in prefix.split('\n')) + len(suffix) + 1 <= max_line_width`